File: infrastructure/cdk/bytesip_stack/bytesip_stack.py
"""ByteSip Infrastructure Stack.

This stack defines the core AWS resources for ByteSip:
- DynamoDB table for news cache
- Lambda function for news fetching
- SSM parameter references for API tokens
"""

from pathlib import Path

from aws_cdk import Duration, RemovalPolicy, Stack
from aws_cdk import aws_dynamodb as dynamodb
from aws_cdk import aws_lambda as lambda_
from aws_cdk import aws_ssm as ssm
from constructs import Construct


class ByteSipStack(Stack):
    """Main infrastructure stack for ByteSip News Agent."""

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        environment: str = "development",
        **kwargs,
    ) -> None:
        """Initialize the ByteSip stack.

        Args:
            scope: CDK scope
            construct_id: Stack ID
            environment: Deployment environment (development, staging, production)
            **kwargs: Additional stack arguments
        """
        super().__init__(scope, construct_id, **kwargs)

        self.deploy_environment = environment

        # DynamoDB table for news cache
        self.news_cache_table = self._create_dynamodb_table()

        # Lambda function for news fetching
        self.news_fetcher_function = self._create_lambda_function()

        # Grant Lambda permissions to access DynamoDB
        self.news_cache_table.grant_read_write_data(self.news_fetcher_function)

    def _create_dynamodb_table(self) -> dynamodb.Table:
        """Create DynamoDB table for news cache.

        Table schema:
        - PK: SOURCE#<source_name> (e.g., SOURCE#qiita)
        - SK: ITEM#<item_id> or META

        Returns:
            DynamoDB Table construct
        """
        table = dynamodb.Table(
            self,
            "NewsCacheTable",
            table_name=f"bytesip-news-cache-{self.deploy_environment}",
            partition_key=dynamodb.Attribute(
                name="PK",
                type=dynamodb.AttributeType.STRING,
            ),
            sort_key=dynamodb.Attribute(
                name="SK",
                type=dynamodb.AttributeType.STRING,
            ),
            billing_mode=dynamodb.BillingMode.PAY_PER_REQUEST,
            time_to_live_attribute="ttl",
            removal_policy=(
                RemovalPolicy.DESTROY
                if self.deploy_environment == "development"
                else RemovalPolicy.RETAIN
            ),
        )
        return table

    def _create_lambda_function(self) -> lambda_.Function:
        """Create Lambda function for news fetching.

        Returns:
            Lambda Function construct
        """
        # Path to Lambda code (relative to CDK project root)
        lambda_code_path = Path(__file__).parent.parent.parent / "lambda" / "bytesip_news_fetcher"

        # SSM parameter names for API tokens
        qiita_token_param = f"/bytesip/{self.deploy_environment}/qiita-access-token"
        github_token_param = f"/bytesip/{self.deploy_environment}/github-access-token"

        function = lambda_.Function(
            self,
            "NewsFetcherFunction",
            function_name=f"bytesip-news-fetcher-{self.deploy_environment}",
            runtime=lambda_.Runtime.PYTHON_3_12,
            handler="handler.lambda_handler",
            code=lambda_.Code.from_asset(str(lambda_code_path)),
            timeout=Duration.seconds(30),
            memory_size=256,
            environment={
                "DYNAMODB_TABLE_NAME": self.news_cache_table.table_name,
                "QIITA_ACCESS_TOKEN_PARAM": qiita_token_param,
                "GITHUB_ACCESS_TOKEN_PARAM": github_token_param,
            },
        )

        # Grant read access to SSM parameters
        qiita_param = ssm.StringParameter.from_secure_string_parameter_attributes(
            self,
            "QiitaTokenParam",
            parameter_name=qiita_token_param,
        )
        github_param = ssm.StringParameter.from_secure_string_parameter_attributes(
            self,
            "GitHubTokenParam",
            parameter_name=github_token_param,
        )
        qiita_param.grant_read(function)
        github_param.grant_read(function)

        return function

File: infrastructure/cdk/lib/bytesip-stack.ts
import * as path from "path";
import * as cdk from "aws-cdk-lib";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as secretsmanager from "aws-cdk-lib/aws-secretsmanager";
import { Construct } from "constructs";

export interface ByteSipStackProps extends cdk.StackProps {
  /**
   * Deployment environment (development, staging, production)
   * @default "development"
   */
  readonly environment?: string;
}

/**
 * ByteSip Infrastructure Stack
 *
 * Creates the core AWS resources for ByteSip News Agent:
 * - DynamoDB table for news cache
 * - Lambda function for news fetching
 * - Secrets Manager for API tokens
 */
export class ByteSipStack extends cdk.Stack {
  public readonly deployEnvironment: string;
  public readonly newsCacheTable: dynamodb.Table;
  public readonly qiitaSecret: secretsmanager.ISecret;
  public readonly githubSecret: secretsmanager.ISecret;
  public readonly newsFetcherFunction: lambda.Function;

  constructor(scope: Construct, id: string, props?: ByteSipStackProps) {
    super(scope, id, props);

    this.deployEnvironment = props?.environment ?? "development";

    // DynamoDB table for news cache
    this.newsCacheTable = this.createDynamoDBTable();

    // Secrets Manager secrets for API tokens
    const secrets = this.createSecrets();
    this.qiitaSecret = secrets.qiitaSecret;
    this.githubSecret = secrets.githubSecret;

    // Lambda function for news fetching
    this.newsFetcherFunction = this.createLambdaFunction();

    // Grant Lambda permissions
    this.newsCacheTable.grantReadWriteData(this.newsFetcherFunction);
    this.qiitaSecret.grantRead(this.newsFetcherFunction);
    this.githubSecret.grantRead(this.newsFetcherFunction);
  }

  private createDynamoDBTable(): dynamodb.Table {
    return new dynamodb.Table(this, "NewsCacheTable", {
      tableName: `bytesip-news-cache-${this.deployEnvironment}`,
      partitionKey: {
        name: "PK",
        type: dynamodb.AttributeType.STRING,
      },
      sortKey: {
        name: "SK",
        type: dynamodb.AttributeType.STRING,
      },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      timeToLiveAttribute: "ttl",
      removalPolicy:
        this.deployEnvironment === "development"
          ? cdk.RemovalPolicy.DESTROY
          : cdk.RemovalPolicy.RETAIN,
    });
  }

  private createSecrets(): {
    qiitaSecret: secretsmanager.ISecret;
    githubSecret: secretsmanager.ISecret;
  } {
    const qiitaToken = process.env.QIITA_ACCESS_TOKEN;
    const githubToken = process.env.GITHUB_ACCESS_TOKEN;

    const qiitaSecretName = `bytesip/${this.deployEnvironment}/qiita-access-token`;
    const githubSecretName = `bytesip/${this.deployEnvironment}/github-access-token`;

    // Only create/update secrets if tokens are provided via environment variables
    // Otherwise, just reference existing secrets
    if (qiitaToken) {
      const qiitaCfnSecret = new secretsmanager.CfnSecret(
        this,
        "QiitaAccessToken",
        {
          name: qiitaSecretName,
          secretString: qiitaToken,
        }
      );
      if (this.deployEnvironment === "development") {
        qiitaCfnSecret.applyRemovalPolicy(cdk.RemovalPolicy.DESTROY);
      }
    }

    if (githubToken) {
      const githubCfnSecret = new secretsmanager.CfnSecret(
        this,
        "GitHubAccessToken",
        {
          name: githubSecretName,
          secretString: githubToken,
        }
      );
      if (this.deployEnvironment === "development") {
        githubCfnSecret.applyRemovalPolicy(cdk.RemovalPolicy.DESTROY);
      }
    }

    // Import secrets for grant operations (assumes secrets exist)
    const qiitaSecret = secretsmanager.Secret.fromSecretNameV2(
      this,
      "QiitaSecretRef",
      qiitaSecretName
    );
    const githubSecret = secretsmanager.Secret.fromSecretNameV2(
      this,
      "GitHubSecretRef",
      githubSecretName
    );

    return { qiitaSecret, githubSecret };
  }

  private createLambdaFunction(): lambda.Function {
    const lambdaBasePath = path.join(__dirname, "..", "..", "lambda");
    const lambdaCodePath = path.join(lambdaBasePath, "bytesip_news_fetcher");

    return new lambda.Function(this, "NewsFetcherFunction", {
      functionName: `bytesip-news-fetcher-${this.deployEnvironment}`,
      runtime: lambda.Runtime.PYTHON_3_12,
      handler: "handler.lambda_handler",
      code: lambda.Code.fromAsset(lambdaBasePath, {
        bundling: {
          image: lambda.Runtime.PYTHON_3_12.bundlingImage,
          command: [
            "bash",
            "-c",
            [
              "pip install requests feedparser -t /asset-output",
              "cp -r bytesip_news_fetcher/* /asset-output/",
            ].join(" && "),
          ],
        },
      }),
      timeout: cdk.Duration.seconds(30),
      memorySize: 256,
      environment: {
        // AWS_REGION is automatically set by Lambda runtime
        DYNAMODB_TABLE_NAME: this.newsCacheTable.tableName,
        QIITA_SECRET_NAME: this.qiitaSecret.secretName,
        GITHUB_SECRET_NAME: this.githubSecret.secretName,
      },
    });
  }
}
